tls_icon.php

```diff
@@ -39,31 +40,32 @@ function get_received_header_content($Received_Header)
 		return $Received;
 	}
 
-	public function storage_init($p)
+	public function message_load($p)
 	{
-		$headers = isset($p['fetch_headers']) ? $p['fetch_headers'] : '';
-		$p['fetch_headers'] = trim(trim($headers) . ' ' . strtoupper('Received'));
-		return $p;
+		// Save message id for retrieving headers
+		$this->message_id = $p['object']->uid;
 	}
 
 	public function message_headers($p)
 	{
 		if ($this->message_headers_done === false) {
 			$this->message_headers_done = true;
 
-			$Received_Header = isset($p['headers']->others['received']) ? $p['headers']->others['received'] : null;
-			$Received = $this->get_received_header_content($Received_Header);
+			// Get raw message headers from storage
+			$headers = $this->rcmail->get_storage()->get_raw_headers($this->message_id);
```

Comment:
Is there any benefit to this flow over the old one? It doesn't look bad, just curious as to whether there is a difference.

Reply:
The new version returns all Received headers instead of only the first one found. The header hook only returns the first of the headers with duplicate keys. 

But that indeed means that you can also hit on the previous hops in the chain. Maybe an extra check for the receiving server (your server) in combination with all the receive headers will do the trick

Reply:
The old code was specifically looking for the `l-n`th header, 0 by default. The config (https://github.com/GermanCoding/Roundcube_TLS_Icon/blob/master/config.inc.php.dist) can be adjusted to configure what `n` should be. 

So it's not true that the hook returns only the first header: For duplicate keys, the Roundcube API returns an array, which we parse and look for a specific user-configurable header (not necessarily the last one, but that's the default).

Reply:
That's weird, when I logged the whole array, only one Received header was returned. But matbe something else went wrong then.